hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudAddressTranslator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.discovery;
+
+import com.hazelcast.client.connection.AddressTranslator;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.LoggingService;
+import com.hazelcast.nio.Address;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class HazelcastCloudAddressTranslator implements AddressTranslator {
+
+    private final ILogger logger;
+    private final HazelcastCloudDiscovery hazelcastCloudDiscovery;
+
+    private volatile Map<Address, Address> privateToPublic = new HashMap<Address, Address>();
+
+    public HazelcastCloudAddressTranslator(String token, LoggingService loggingService) {
+        this(new HazelcastCloudDiscovery(token), loggingService);
+    }
+
+    // just for testing
+    HazelcastCloudAddressTranslator(HazelcastCloudDiscovery hazelcastCloudDiscovery, LoggingService loggingService) {
+        this.hazelcastCloudDiscovery = hazelcastCloudDiscovery;
+        this.logger = loggingService.getLogger(HazelcastCloudAddressTranslator.class);
+    }
+
+    @Override
+    public Address translate(Address address) {
+        if (address == null) {
+            return null;
+        }
+        Address publicAddress = privateToPublic.get(address);
+        if (publicAddress != null) {
+            return publicAddress;
+        }
+
+        refresh();
```

Comment:
This may cause 30 sec delay. Assume that as a user I configured an Address as "unknown host:5555" in the ClientConfig for which there will never be such a host, then this code will always execute each time the cluster is being connected by the client and this address is being tried, right? How good is this behaviour?

Reply:
I understand the concern. But what is the solution ?
I can't just blacklist the addresses that are not found. Because they might be valid but not reported yet. 

This made me think. Should we disallow giving addresses via config(xml/programmetic) when a discovery enabled ? 
May be `isDiscoveryConfigurationConsistent` should also check for that. 
Wdyt ?